### ISSUE TITLE
A *hidden* interactive element inside a <button> is flagged as an issue - should it be?

### ISSUE THREAD

**Issue by lloydi** [validator/validator#1188]
URL being validated or code to reproduce error:
N/A internal system, not public

Validator says:

Error: The element `input` must not appear as a descendant of an element with the attribute `role=button`.

```
<div role="button" tabindex="0" aria-labelledby="something">
<input id="theThing" type="file" style="display:none">
<div>
```

This is a pattern where a default file upload input has been hidden for a better designed file upload button.

On the basis that the element being flagged is hidden (completely, not just visually), and therefore cannot be accessed, should this be flagged as an issue? Or does the hidden state have no bearing on it (because, potentially, an element's state might change dynamically)?

**Comment by sideshowbarker** [comment]
Short answer: It should be flagged — and so, the checker is behaving as expected — because the spec requirements don’t treat the case of an interactive element with `style="display:none"` any differently than they do the case of an interactive element without that.

In more detail:

At https://html.spec.whatwg.org/multipage/form-elements.html#the-button-element:interactive-content-2-2, the spec says this about buttons:

> there must be no [interactive content](https://html.spec.whatwg.org/multipage/dom.html#interactive-content-2) descendant

And _interactive content_ is defined at https://html.spec.whatwg.org/multipage/dom.html#interactive-content-2 as:

>`a` (if the `href` attribute is present), `audio` (if the `controls` attribute is present), `button`, `details`, `embed`, `iframe`, `img` (if the `usemap` attribute is present), `input` (if the `type` attribute is not in the Hidden state), `label`, `select`, `textarea`, `video` (if the `controls` attribute is present)

Specifically, in neither of those places does the spec qualify any of those statements with any requirement about whether or not the element is un-displayed/hidden, or whatever.
